2018/003-offline-chart-repository.md

```diff
@@ -0,0 +1,441 @@
+-rw-r--r-- 1 flavio users  3873 Jul  3 10:00 console-1.0.2.tgz
+-rw-r--r-- 1 flavio users  4078 Jul  3 10:00 console-1.1.0.tgz
+-rw-r--r-- 1 flavio users 11243 Jul  3 10:00 uaa-2.0.2.tgz
+-rw-r--r-- 1 flavio users  6180 Jul  3 10:00 uaa-2.10.1.tgz
+-rw-r--r-- 1 flavio users  6988 Jul  3 10:00 uaa-2.4.1-beta3+cf278.0.g12f5d3a8.tgz
+-rw-r--r-- 1 flavio users  4152 Jul  3 10:00 uaa-2.5.0-beta4+cf278.0.gafa3d0e9.tgz
+-rw-r--r-- 1 flavio users  4143 Jul  3 10:00 uaa-2.6.11.tgz
+-rw-r--r-- 1 flavio users  4175 Jul  3 10:00 uaa-2.6.1-rc1+cf278.0.g52d7a644.tgz
+-rw-r--r-- 1 flavio users  4144 Jul  3 10:00 uaa-2.7.0.tgz
+-rw-r--r-- 1 flavio users  5470 Jul  3 10:00 uaa-2.8.0.tgz
+```
+
+A simple read-only chart registry can be created with the following command:
+
+```
+docker run --rm \
+  -v /home/flavio/charts:/charts:ro \
+  -p 4000:4000 \
+  chartmuseum/chartmuseum \
+    --port 4000 \
+    --disable-api \
+    --storage local \
+    --storage-local-rootdir /charts
+```
+The above command mounts the charts directory from the host into the container
+using the read-only mode. Then it starts ChartMuseum with the following flags:
+
+* `--disable-api`: all the write operations (create, update, delete) are
+  disabled, only read operations are allowed.
+* `--storage` and `--storage-local-rootdir`: all the charts are stored on the
+  local filesystem and are inside of the `/charts` directory.
+
+The `index.yaml` file can be obtained by executing: `curl http://localhost:4000`.
+The file is generated dynamically by ChartMuseum at boot time and is kept
+in memory.
+
+#### The offline chart repository image
+
+SUSE already plans to distribute a ChartMuseum image to all its customers. The
+purpose of this is image is to allow them to run their own internal chart
+repository where they can upload their personal charts.
+
+To accomplish that SUSE will create and maintain a container image of
+ChartMuseum. During the rest of this RFC we will address this image as
+`registry.suse.com/chartmuseum`.
+
+The [entrypoint](https://docs.docker.com/engine/reference/builder/#entrypoint)
+of this image will start ChartMuseum in "normal" mode, meaning all the API endpoints will be enabled (making the repository read-write).
+
+
+The SUSE chart repository available at [kubernetes-charts](https://kubernetes-charts.suse.com)
+is going to be shipped inside of an image called
+`registry.suse.com/suse/kubernetes-charts` (again, the name is just an example
+chosen for this RFC, the final name might change).
+
+The `registry.suse.com/suse/kubernetes-charts` image is going to be based
+on the `registry.suse.com/chartmuseum` image. It will extend it by:
+
+  * Adding the RPM packages providing the helm charts of SUSE products
+  * Providing a new entrypoint that makes the registry read-only and points
+    ChartMuseum to the location where all the SUSE charts have been
+    installed by the RPM.
+
+Ideally each SUSE team will ship their own charts via a single RPM per chart. All the
```

Comment:
This is a requirement. Multiple folders are not allowed. But you can create sub folders to create a "namespaced" setup.